### PR TITLE
enhance FloatHistogram CopyToSchema method

### LINKED ISSUE #11270: histograms: Possible FloatHistogram.CopyToSchema performance issue
Vet the performance impact of the `FloatHistogram.CopyToSchema` method and optimize code if needed (see TODO in the code).

### PR DESCRIPTION
**Pull request by fatsheep9146**
Fix #11270

The basic idea behind the current implementation of the CopyToSchema method is

- first convert the `FloatHistogram.PositiveSpans/NegativeSpans` into targetSchema  by `floatBucketIterator`
- then got each bucket in targetSchema by `Next()`
- then add back to the target `FloatHistogram` object

Basically we can summarize the process is OriginalSpan[] -> TargetBucket[] -> TargetSpan[]

So I think we could enhance this process into OriginalSpan[] -> TargetSpan, by directly processing the Spans

So I wrote a simple implementation in this idea, with Function name `CopyToSchemaNew` and wrote a benchmark test for `CopyToSchema` and `CopyToSchemaNew`

The test result shows that the enhanced one is 50% faster than orignal one on my laptop

```
BenchmarkCopyToSchema-8      	 1176043	      1023 ns/op
BenchmarkCopyToSchemaNew-8   	 2502216	       498.0 ns/op
```

### CODE REVIEW

File: model/histogram/float_histogram.go [hunk 1]
```diff
@@ -982,3 +964,72 @@ func (i *allFloatBucketIterator) Next() bool {
...
 }
+
+func targetIdx(idx, schema, targetSchema int32) int32 {
+	return ((idx - 1) >> (schema - targetSchema)) + 1
+}
+
+func convertSpanToSchema(spans []Span, counts []float64, schema, targetSchema int32) ([]Span, []float64) {
+	targetSpans := make([]Span, 0)
+	targetCounts := make([]float64, 0)
+	// the index of bucket for origin schema
```
Comment (beorn7) [inline model/histogram/float_histogram.go:975]: Generally, we aim for easily readable English in comments throughout the Prometheus codebase. So:
- Start with a capital letter.
- End with a period.
- Don't leave out syntactically required words for brevity.

So this should read:
```Go
// The index of a bucket in the original schema.
```

This is probably still short enough to go on the same line, especially if using the `var` grouping suggested below:
```Go
var (
    // ...
    bucketIdx int32 // The index of a bucket in the original schema.
    // ...
)
```

Same applies to the other code comments below.

File: model/histogram/float_histogram.go [hunk 2]
```diff
@@ -982,3 +964,72 @@ func (i *allFloatBucketIterator) Next() bool {
...
+	var bucketIdx int32 = 0
+	var lastTargetBucketIdx int32 = 0
+	// the index of bucket in a span
```
Comment (beorn7) [inline model/histogram/float_histogram.go:978]: That's not quite right, is it? It appears to me this is used for the position of a bucket in the original buckets slice (currently called `counts`, but I would propose `buckets` or `origBuckets`).

File: model/histogram/float_histogram.go [hunk 3]
```diff
@@ -982,3 +964,72 @@ func (i *allFloatBucketIterator) Next() bool {
...
+
+			switch {
+			case len(targetSpans) == 0:
+				// this is the first span
+				span := Span{
+					Offset: targetBucketIdx,
+					Length: 1,
+				}
+				targetSpans = append(targetSpans, span)
+				targetCounts = append(targetCounts, counts[spanBucketIdx])
```
Comment (beorn7) [inline model/histogram/float_histogram.go:1001]: `spanBucketIdx` must be 0 here, so I'd use 0 explicitly for clarity.

File: model/histogram/float_histogram.go [hunk 4]
```diff
@@ -982,3 +964,72 @@ func (i *allFloatBucketIterator) Next() bool {
...
+
+	for i, span := range spans {
+		// determine the boundary of first bucket in this span
+		if i == 0 {
+			bucketIdx = spans[0].Offset
+		} else {
+			bucketIdx += span.Offset
+		}
+		// iterate over each span
+		for j := 0; j < int(span.Length); j, bucketIdx = j+1, bucketIdx+1 {
```
Comment (beorn7) [inline model/histogram/float_histogram.go:989]: I would write `for j := 0; j < int(span.Length); j++ {` and increment `bucketIdx` together with `spanBucketIdx` below.

File: model/histogram/float_histogram.go [hunk 5]
```diff
@@ -982,3 +964,72 @@ func (i *allFloatBucketIterator) Next() bool {
...
+					Offset: targetBucketIdx,
+					Length: 1,
+				}
+				targetSpans = append(targetSpans, span)
+				targetCounts = append(targetCounts, counts[spanBucketIdx])
+				lastTargetBucketIdx = targetBucketIdx
+
+			case lastTargetBucketIdx == targetBucketIdx:
+				// the target bucket just locates current last span,
+				// then increase counter of last span
```
Comment (beorn7) [inline model/histogram/float_histogram.go:1006]: This reads a bit weird. How about: "The current bucket has to be merged into the same target bucket as the previous bucket."

File: model/histogram/float_histogram.go [hunk 6]
```diff
@@ -982,3 +964,72 @@ func (i *allFloatBucketIterator) Next() bool {
...
+				targetCounts[len(targetCounts)-1] += counts[spanBucketIdx]
+
+			case (lastTargetBucketIdx + 1) == targetBucketIdx:
+				// the target bucket just locates next to current last span,
+				// extend last span
```
Comment (beorn7) [inline model/histogram/float_histogram.go:1011]: How about: "The current bucket has to go into a new target bucket, and that bucket is next to the previous target bucket, so we add it to the current target span."

File: model/histogram/float_histogram.go [hunk 7]
```diff
@@ -982,3 +964,72 @@ func (i *allFloatBucketIterator) Next() bool {
...
+				// create new span
+				span := Span{
+					Offset: targetBucketIdx - lastTargetBucketIdx - 1,
+					Length: 1,
+				}
+				targetSpans = append(targetSpans, span)
+				targetCounts = append(targetCounts, counts[spanBucketIdx])
+				lastTargetBucketIdx = targetBucketIdx
+
+			default:
```
Comment (beorn7) [inline model/histogram/float_histogram.go:1027]: No empty `default` needed.

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thanks for doing this. I think the code is correct. My comments are only about style (naming, comments, code).

File: model/histogram/float_histogram.go [hunk 8]
```diff
@@ -982,3 +964,72 @@ func (i *allFloatBucketIterator) Next() bool {
...
+	// the index of bucket in a span
+	spanBucketIdx := 0
+
+	for i, span := range spans {
+		// determine the boundary of first bucket in this span
+		if i == 0 {
+			bucketIdx = spans[0].Offset
+		} else {
+			bucketIdx += span.Offset
+		}
```
[the remark this answers is not in the document]
Comment (fatsheep9146) [reply model/histogram/float_histogram.go:987]: Yes, I also noticed this, but I still first write like this to make sure the logic is easy to be understood. I will change it as you said.

**Comment by fatsheep9146** [comment]
All comments are resolved
@beorn7

File: model/histogram/float_histogram.go [hunk 9]
```diff
@@ -982,3 +964,71 @@ func (i *allFloatBucketIterator) Next() bool {
...
 }
+
+// targetIdx returns the bucket index in the target schema for the given bucket
+// index idx in the original schema.
+func targetIdx(idx, originSchema, targetSchema int32) int32 {
+	return ((idx - 1) >> (originSchema - targetSchema)) + 1
+}
+
+// mergeToSchema is used to merge a FloatHistogram's Spans and Buckets (no matter positive or negative) from original schema to target schema
+// the target schema must smaller than original schema
```
Comment (beorn7) [inline model/histogram/float_histogram.go:975]:
```suggestion
// mergeToSchema is used to merge a FloatHistogram's Spans and Buckets (no matter if 
// positive or negative) from the original schema to the target schema.
// The target schema must be smaller than the original schema.
```

File: model/histogram/float_histogram.go [hunk 10]
```diff
@@ -982,3 +964,71 @@ func (i *allFloatBucketIterator) Next() bool {
...
+func mergeToSchema(originSpans []Span, originBuckets []float64, originSchema, targetSchema int32) ([]Span, []float64) {
+	var (
+		targetSpans         []Span    // The spans in the target schema.
+		targetBuckets       []float64 // The buckets in the target schema.
+		bucketIdx           int32     // The index of bucket in the origin schema.
+		lastTargetBucketIdx int32     // The index of bucket which is the last bucket for current targetSpans and targetBuckets.
```
Comment (beorn7) [inline model/histogram/float_histogram.go:981]: Maybe just "The index of the last added target bucket." ?

**Review by beorn7 — CHANGES_REQUESTED** [review]
Beautiful. Just two more comment tweaks.

File: model/histogram/float_histogram.go (hunk 10 above)
Comment (fatsheep9146) [reply model/histogram/float_histogram.go]: Ok! Done

**Comment by fatsheep9146** [comment]
@beorn7 All comments are fixed.

**Pull request merged**